js_modules/dagster-ui/packages/ui-core/src/app/Util.tsx

```diff
@@ -148,6 +148,53 @@ export function asyncMemoize<T, R, U extends (arg: T, ...rest: any[]) => Promise
   }) as any;
 }
 
+export function localStorageAsyncMemoize<
+  T,
+  R,
+  U extends (localStorageKey: string, versionKey: string, arg: T, ...rest: any[]) => Promise<R>,
+>(
+  // Key to that acts as a DB for memoized values
+  fn: U,
+  hashFn?: (arg: T, ...rest: any[]) => any,
+): U {
+  return (async (
+    // Allows for independent db invalidation on per code location + group basis
+    localStorageKey: string,
+    // Key that is used to determine if a value in the DB is stale
+    versionKey: string,
+    arg: T,
+    ...rest: any[]
+  ) => {
+    const key = `localStorageAsyncMemoize:${localStorageKey}`;
+    type DBWithVersion = Record<string, R> & {version: string};
+    const hashKey = hashFn ? hashFn(arg, ...rest) : arg;
+    let db = {version: versionKey} as DBWithVersion;
+    const dbString = localStorage.getItem(key);
+    if (dbString) {
+      try {
+        db = JSON.parse(dbString!);
+      } catch (e) {
+        localStorage.removeItem(key);
+        db = {version: versionKey} as DBWithVersion;
+      }
+    }
+    if (db && db.version === versionKey) {
+      if (db[hashKey]) {
+        return db[hashKey];
+      }
+    } else {
+      // If the version has changed, clear the DB
+      localStorage.removeItem(key);
+      db = {version: versionKey} as DBWithVersion;
+    }
+
+    const r = (await fn(localStorageKey, versionKey, arg, ...rest)) as R;
+    db[hashKey] = r;
+    localStorage.setItem(key, JSON.stringify(db));
```

Comment:
This is super cool but I think we need to think more about the cache invalidation policy to avoid filling local storage. Our whole site only gets 5MB and setItem throws exceptions once you exceed it, which could result in weird crashes in other places that assume they can store a small key without checking for exceptions...

I think we just need a way to avoid the set of hashKeys growing forever in the `db` state. Maybe we implement a very small LRU cache and only keep the last 5 cache hits/misses?

It's probably possible that a single layout object could serialize to >5MB,  and we probably need to assume that this `setItem` can throw an exception and still have everything "work" the old way.

Reply:
See my comment below but I don't think 5MB is the actual limit anymore.